<a href="https://colab.research.google.com/github/Kadeian/reimagined-octo-disco/blob/master/Copy_of_CorrectedJayspickling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import sklearn
import sklearn.datasets
import sklearn.ensemble
import sklearn.model_selection

In [ ]:
iris = sklearn.datasets.load_iris()

In [ ]:
train, test, labels_train, labels_test = sklearn.model_selection.train_test_split(iris.data, iris.target, train_size=0.80)

In [ ]:
rf = sklearn.ensemble.RandomForestClassifier(n_estimators=500)
rf.fit(train, labels_train)
rf

RandomForestClassifier(n_estimators=500)

In [ ]:
import sklearn
import sklearn.datasets
import sklearn.ensemble
import sklearn.model_selection
import joblib  # Updated import statement
import os
import pickle

# Load data
iris = sklearn.datasets.load_iris()
train, test, labels_train, labels_test = sklearn.model_selection.train_test_split(iris.data, iris.target, train_size=0.80)

# Train model
rf = sklearn.ensemble.RandomForestClassifier(n_estimators=500)
rf.fit(train, labels_train)

# Save and load model using pickle
filename_pickle = 'sunil.pras'
with open(filename_pickle, 'wb') as file:
    pickle.dump(rf, file)

with open(filename_pickle, 'rb') as file:
    loaded_model = pickle.load(file)
result_pickle = loaded_model.score(test, labels_test)
print('Pickle Result:', result_pickle)

# Save and load model using joblib
filename_joblib = 'finalized_model_joblib.sav'
joblib.dump(rf, filename_joblib)
loaded_model_joblib = joblib.load(filename_joblib)
result_joblib = loaded_model_joblib.score(test, labels_test)
print('Joblib Result:', result_joblib)



Pickle Result: 0.9666666666666667
Joblib Result: 0.9666666666666667


# Machine Learning Workflow Explanation

Let's break down each part of what you've done, including the terminology and the processes involved in your machine learning workflow:

## 1. Creating and Training a Model:
- **Data Loading (`iris = sklearn.datasets.load_iris()`)**: You loaded the Iris dataset, which is a famous dataset containing data on iris flowers. It includes features like petal length, petal width, sepal length, and sepal width, and the target variable is the species of the iris.
- **Data Splitting**: You split the dataset into training and testing sets. 80% of the data is used for training (`train`), where the model learns the patterns, and 20% is used for testing (`test`), where the model is evaluated.
- **Model Training (`rf.fit(train, labels_train)`)**: You trained a RandomForestClassifier. `rf` stands for "Random Forest." It is an ensemble learning method for classification (and regression) that works by constructing a multitude of decision trees at training time and outputting the class that is the mode of the classes (classification) of the individual trees.

## 2. Saving and Loading the Model:
- **`pickle.dump()` and `pickle.load()`**: These functions are used to serialize and deserialize a Python object structure. "Serialization" refers to the process of converting an object's state to a format that can be stored (e.g., in a file) or transmitted (e.g., across a network). Deserialization is the opposite—converting the serialized format back into an object.
- **`'wb'` and `'rb'`**: These are modes used when opening a file. `'wb'` means "write binary," used for writing to files in binary mode (as opposed to text mode). `'rb'` means "read binary," used for reading from files in binary mode. Binary mode is used here because the model structure involves binary data, which is more complex than just text.

## 3. Using `joblib` for Serialization:
- **`joblib.dump()` and `joblib.load()`**: `joblib` is particularly suited for models that involve large arrays of data, as is typical in scikit-learn. This library is optimized for these use cases to be more efficient than `pickle`.
- **Saving and loading with `joblib`**: This follows the same principle as `pickle`. You save the model to disk and then load it back to predict or validate it on new data.

## 4. Evaluating the Model:
- **`score()` Method**: After loading the model (either through `pickle` or `joblib`), you used its `score()` method to determine the accuracy of the model on the test dataset. This method computes the mean accuracy, comparing the predicted labels with the true labels.

## 5. Results:
- **Printed Accuracy**: The accuracy scores from both the `pickle` and `joblib` loaded models were approximately 96.67%, which indicates how well the models predicted the test data.

This workflow you completed is a common pattern in data science and machine learning for creating, saving, loading, and evaluating models. It encapsulates the entire process of model development, including validation and deployment phases. If you have more questions or need further explanations on any of these steps, feel free to ask!
